### ISSUE TITLE
Opt-in warn when using iter()-like methods on Option or Result

### ISSUE THREAD

**Issue by kangalioo** [rust-lang/rust-clippy#6762]
### What it does

Warn when iter, iter_mut or into_iter is used on an Option or Result.

### Categories (optional)

- Kind: clippy::pedantic

*What is the advantage of the recommended code over the original code*

Treating Option/Result as normal iterables is sometimes useful, but also sometimes produces really confusing behavior, especially on types like `Option<Vec<_>>`. See below for an example where a hidden bug gets uncovered by removing the obfuscation caused by treating Option/Result as an iterable.

### Drawbacks

Treating Option/Result as iterables is sometimes useful, for example in long clever adapter chains. Personally I still think it's a code smell to do that, but to each their own.

### Example

```rust
pub fn get_item_list() -> Result<Vec<Item>, Error> { /* ... */ }

// Confusing, isn't it?
for item in get_item_list() {
    // ...
}
```

Could be written as:

```rust
pub fn get_item_list() -> Result<Vec<Item>, Error> { /* ... */ }

// Now it's obvious what's happening
if let Ok(item) = get_item_list() {
    // ...
}
```

**Comment by camsteffen** [comment]
For loops, as in your example, are already covered by `for_loops_over_fallibles` and deny by default.

**Comment by kangalioo** [comment]
Ah my bad. I had tried to search for an existing lint but I was using the wrong search terms.

This lint can still be useful though, less for code correctness than for code quality

**Comment by camsteffen** [comment]
At a minimum we should lint `for _ in opt.iter()` or `..into_iter()`. That's a false negative.

**Comment by fujiapple852** [comment]
Somewhat related, I just saw a (now deleted) question on SO where the submitter had a bug which `for_loops_over_fallibles` would have detected, had they run Clippy that is.  

However the ["Why is this bad?"](https://rust-lang.github.io/rust-clippy/master/index.html#for_loops_over_fallibles) description for this lint suggests, with a casual reading, that this is _just_ a readability issue and doesn't mention the "treating Option/Result as an iterable" gotcha at all:

> Readability. This is more clearly expressed as an if let.

Perhaps the this lint could benefit with a clearer explanation and perhaps a references to [Result::iter](https://doc.rust-lang.org/std/result/enum.Result.html#method.iter), [Option::iter](https://doc.rust-lang.org/std/option/enum.Option.html#method.iter) and documentation describing the relationship between [for loops and IntoIterator](https://doc.rust-lang.org/std/iter/index.html#for-loops-and-intoiterator)?
